### PR TITLE
Start/stop ProcessStateMonitor with ApplicationStateTracker

### PR DESCRIPTION
**Pull request by szewai**
#### d5a19d183b733b93b0a202ea65651a2614200998
<pre>
Start/stop ProcessStateMonitor with ApplicationStateTracker
<a href="https://bugs.webkit.org/show_bug.cgi?id=241427">https://bugs.webkit.org/show_bug.cgi?id=241427</a>

Reviewed by Geoffrey Garen.

We want ProcessStateMonitor to be started when application is in background and to be stopped when application is in
foreground. In r294405, we start/stop ProcessStateMonitor on receiving UIApplicationDidEnterBackgroundNotification and
UIApplicationWillEnterForegroundNotification notifications. However, UIApplicationWillEnterForegroundNotification can be
sent later than UISceneWillEnterForegroundNotification. In ApplicationStateTracke, we start creating foreground activity
for web process on receiving UISceneWillEnterForegroundNotification. At this time,
UIApplicationWillEnterForegroundNotification may not be received yet and ProcessStateMonitor may still be
working. ProcessStateMonitor may forbid creating new activity after it sets process shouldSuspend (see r294405), so we
need to make sure ProcessStateMonitor is stopped before starting new foreground activity.

To achieve that, we now decide when to start/stop ProcessStateMonitor with ApplicationStateTracker. We start
ProcessStateMonitor when all trackers are in the background, and stop it when at least one tracker is in foreground.

* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(WebKit::allApplicationStateTrackers):
(WebKit::updateApplicationBackgroundState):
(WebKit::ApplicationStateTracker::ApplicationStateTracker):
(WebKit::ApplicationStateTracker::~ApplicationStateTracker):
(WebKit::ApplicationStateTracker::applicationDidEnterBackground):
(WebKit::ApplicationStateTracker::applicationWillEnterForeground):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::setProcessesShouldSuspend):
* Source/WebKit/UIProcess/ProcessAssertion.h:
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/ProcessAssertionIOS.mm:
(-[WKProcessAssertionBackgroundTaskManager _releaseBackgroundTask]):
(-[WKProcessAssertionBackgroundTaskManager setProcessStateMonitorEnabled:]):
(WebKit::ProcessAndUIAssertion::setProcessStateMonitorEnabled):
* Source/WebKit/UIProcess/ios/ProcessStateMonitor.mm:
(WebKit::ProcessStateMonitor::checkRemainingRunTime):
* Source/WebKit/UIProcess/ios/WebProcessProxyIOS.mm:
(WebKit::WebProcessProxy::platformInitialize):

Canonical link: <a href="https://commits.webkit.org/251498@main">https://commits.webkit.org/251498@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295493">https://svn.webkit.org/repository/webkit/trunk@295493</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm [hunk 1]
```diff
@@ -50,6 +51,28 @@ - (BOOL)_isHostedInAnotherProcess;
...
+{
+    static NeverDestroyed<WeakHashSet<ApplicationStateTracker>> trackers;
+    return trackers;
+}
+
+static void updateApplicationBackgroundState()
+{
+    static bool isInBackground = false;
+    auto isAnyStateTrackerInBackground = []() -> bool {
+        return WTF::allOf(allApplicationStateTrackers(), [](auto& tracker) {
```
Comment (geoffreygaren) [inline Source/WebKit/UIProcess/ApplicationStateTracker.mm]: Don't you mean anyOf() here?

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm [hunk 2]
```diff
@@ -50,6 +51,28 @@ - (BOOL)_isHostedInAnotherProcess;
...
+static WeakHashSet<ApplicationStateTracker>& allApplicationStateTrackers()
+{
+    static NeverDestroyed<WeakHashSet<ApplicationStateTracker>> trackers;
+    return trackers;
+}
+
+static void updateApplicationBackgroundState()
+{
+    static bool isInBackground = false;
+    auto isAnyStateTrackerInBackground = []() -> bool {
```
Comment (geoffreygaren) [inline Source/WebKit/UIProcess/ApplicationStateTracker.mm]: Or maybe here you mean to say "areAllStateTrackersInBackground"?

**Review by geoffreygaren — COMMENTED** [review]
> ProcessStateMonitor may forbid creating new activity and taking new assertion (see r294405)

Can you clarify, is this the intentional behavior of ProcessStateMonitor, or an unwanted side effect from something?

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm (hunk 1 above)
Comment (szewai) [reply Source/WebKit/UIProcess/ApplicationStateTracker.mm]: No, should be changed to areAlllStateTrackersInBackground.

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm (hunk 2 above)
Comment (szewai) [reply Source/WebKit/UIProcess/ApplicationStateTracker.mm]: Yes.

**Comment by szewai** [comment]
> > ProcessStateMonitor may forbid creating new activity and taking new assertion (see r294405)
> 
> Can you clarify, is this the intentional behavior of ProcessStateMonitor, or an unwanted side effect from something?

This is intentional behavior: ProcessStateMonitor will mark process as suspended (which means process should not start any new activity) when process's remaining runtime is less than 15s (or process has given up assertions).

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm (hunk 2 above)
Comment (geoffreygaren) [reply Source/WebKit/UIProcess/ApplicationStateTracker.mm]: FWIW, if can arrange to use anyOf() and a not operator, that's a little faster. This is not terribly important.

Comment (szewai) [reply Source/WebKit/UIProcess/ApplicationStateTracker.mm]: Done.

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm [hunk 3]
```diff
@@ -50,6 +51,28 @@ - (BOOL)_isHostedInAnotherProcess;
...
+static WeakHashSet<ApplicationStateTracker>& allApplicationStateTrackers()
+{
+    static NeverDestroyed<WeakHashSet<ApplicationStateTracker>> trackers;
+    return trackers;
+}
+
+static void updateApplicationBackgroundState()
+{
+    static bool isApplicationInBackground = false;
+    auto isAnyStateTrackersInForeground = []() -> bool {
```
Comment (geoffreygaren) [inline Source/WebKit/UIProcess/ApplicationStateTracker.mm]: "is" is singular, so English grammar prefers "isAnyStateTrackerInForeground".

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm [hunk 4]
```diff
@@ -50,6 +51,28 @@ - (BOOL)_isHostedInAnotherProcess;
...
 
+static WeakHashSet<ApplicationStateTracker>& allApplicationStateTrackers()
+{
+    static NeverDestroyed<WeakHashSet<ApplicationStateTracker>> trackers;
+    return trackers;
+}
+
+static void updateApplicationBackgroundState()
+{
+    static bool isApplicationInBackground = false;
```
Comment (geoffreygaren) [inline Source/WebKit/UIProcess/ApplicationStateTracker.mm]: WebKit style puts "s_" in the front of a static to call out its global status.

File: Source/WebKit/UIProcess/ApplicationStateTracker.mm [hunk 5]
```diff
@@ -50,6 +51,28 @@ - (BOOL)_isHostedInAnotherProcess;
...
+    auto isAnyStateTrackersInForeground = []() -> bool {
+        return WTF::anyOf(allApplicationStateTrackers(), [](auto& tracker) {
+            return !tracker.isInBackground();
+        });
+    };
+    bool currentIsApplicationInBackground = !isAnyStateTrackersInForeground();
```
Comment (geoffreygaren) [inline Source/WebKit/UIProcess/ApplicationStateTracker.mm]: "current" can be ambiguous at a time of change like this.

If you switch to "s_" then you can just call this isApplicationInBackground.

**Review by geoffreygaren — APPROVED** [review]
r=me

I made some new naming suggestions; OK to apply them and land without another review.

**Pull request merged**